README.rst
============
``gs.cache``
============
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
A shared cache between processes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Author: `Michael JasonSmith`_
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2014-04-25
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.Net`_.

Introduction
============

A GroupServer system may be served by multiple threads running in
multiple Zope processes; having an external *shared* cache allows
every thread maintain a consistent state. The ``gs.cache``
product provides this caching mechanism. Behind the scenes `the
Python redis module`_ is used to store the key-value pairs in
`the Redis data structure server`_. If Redis is unavailable a
null-cache is used.

The `gs.cache.cache`_ decorator is the main mechanism for using
the cache. Simpler non-persistent per-thread caches can be
created using ``zope.cachedescriptors`` [#zcd]_.

.. _the Python redis module: http://pypi.python.org/pypi/redis/
.. _the redis data structure server: http://redis.io/

``gs.cache.cache``
==================

Lazily evaluate a function, storing the result in a shared cache.

Synopsis
--------

::

  @gs.cache.cache(cacheName, cacheKeyFunc, [expiry])
  def some_function(…):

Description
-----------

The ``gs.cache.cache`` function decorator adds multi-thread
caching to a function. The ``cacheKeyFunction`` is evaluated to
generate a key. Then the cache (possibly stored using Redis)
specified by ``cacheName`` is checked for the key. If present the
value stored in the cache is returned. Otherwise the decorated
function is evaluated, the returned value stored in the cache
(under the cache key), and returned.

Arguments
---------

``cacheName``:
  The name of the cache to store the results in (``str``).

``cacheKeyFunc``:
  A function to generate the cache-key. It will be passed the
  same arguments ast the decorated function.

``expiry``:
  The optional expiry date, in seconds, for items in the cache
  (``int``).

Returns
-------

* The value of the decorated function if the cache-key returned
  by ``cacheKeyFunc`` is absent from the cache specified by
  ``cacheName``.
* The value pointed to by ``cacheKeyFunc`` in the cache specified
  by ``cacheName`` if there is a value stored in the cache.

Side effects
------------

On a cache-miss the decorated function is evaluated. The value
returned by the decorated function is stored in the cache under
the value specified by the cache-key.

Example
-------

The example below [#list]_ shows the use of ``gs.cache.cache``
decorator. The function ``gs.dmarc.lookup_receiver_policy`` is
slow, because it has to make a DNS request to retrieve the DMARC
policy for a site [#dmarc]_. To mitigate this issue a cache is
added::

    @gs.cache.cache('Products.XWFMailingList.dmarc', lambda s, h: h, 7 * 60)
    def get_dmarc_policy_for_host(self, host):
        retval = gs.dmarc.lookup_receiver_policy(host)
        return retval

To add a cache the original ``lookup_receiver_policy`` function
is wrapped in ``get_dmarc_policy_for_host`` [#static]_. This
function is decorated with ``gs.cache.cache``.

* The cache-name is ``Products.XWFMailingList.dmarc``.
* The function to generate the cache-key is a lambda-function. It
  takes the same two arguments as the wrapped function (``self``,
  and ``host``) and returns the second as the key (``host``).
* Because DMARC policy settings can change a cache-timeout of
  seven minutes (``7 * 60`` seconds) is set.

Resources
=========

- Code repository: https://source.iopen.net/groupserver/gs.cache/
- Questions and comments to http://groupserver.org/groups/development
- Report bugs at https://redmine.iopen.net/projects/groupserver

.. [#zcd] For more information on ``zope.cachedescriptors`` see
          <https://pypi.python.org/pypi/zope.cachedescriptors>
.. [#list] The example is taken from
           ``Products.XWFMailingListManager.XWFMailingList``.
.. [#dmarc] For more information on ``gs.dmarc`` see
            <https://source.iopen.net/groupserver/gs.dmarc/>.
.. [#static] The **actual** ``get_dmarc_policy_for_host`` method
             is a static-method. This results in a
             double-decorator, so the function declaration
             actually looks like the following::

               @staticmethod
               @cache('Products.XWFMailingList.dmarc', lambda h: h, 7 * 60)
               def get_dmarc_policy_for_host(host):

.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net
.. _Michael JasonSmith: http://groupserver.org/p/mpj17
.. _Creative Commons Attribution-Share Alike 4.0 International License:
    http://creativecommons.org/licenses/by-sa/4.0/

..  LocalWords:  redis


docs/HISTORY.txt
Changelog
=========

2.1.0 (2014-04-25)
------------------

* Added Python 3 support
* Updated and added to unit tests
* Added tox support
* Refactor into multiple modules

2.0.0 (2013-08-09)
------------------

* PEP-8 code cleanup
* New caching backends.

1.0.0 (2012-06-22)
------------------

* Initial version


gs/cache/tests/backends.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
from __future__ import absolute_import, unicode_literals
import sys
if (sys.version_info >= (3, )):
    from pickle import dumps
else:
    from cPickle import dumps  # lint:ok
from unittest import TestCase
from mock import MagicMock
import gs.cache.backends


class TestRedis(TestCase):

    def setUp(self):
        b = MagicMock()
        self.cacheName = 'test-cache'
        self.redisBackend = gs.cache.backends.RedisCache(b, self.cacheName)

    def test_set(self):
        key = 'Ethyl'
        val = 'The Frog'
        self.redisBackend.backend.set = MagicMock()
        self.redisBackend.set(key, val)
        self.assertEqual(1, self.redisBackend.backend.set.call_count)
        args, kw_args = self.redisBackend.backend.set.call_args
        self.assertIn(key, args[0])
        self.assertIn(self.cacheName, args[0])
        self.assertEqual(dumps(val), args[1])

    def test_has_key(self):
        key = 'Ethyl'
        self.redisBackend.backend.keys = MagicMock(return_value=True)
        self.redisBackend.has_key(key)  # lint:ok
        self.assertEqual(1, self.redisBackend.backend.keys.call_count)
        args, kw_args = self.redisBackend.backend.keys.call_args
        self.assertIn(key, args[0])
        self.assertIn(self.cacheName, args[0])

    def test_get(self):
        key = 'Ethyl'
        val = 'The Frog'
        self.redisBackend.backend.get = MagicMock(return_value=dumps(val))
        r = self.redisBackend.get(key)
        self.assertEqual(val, r)
        self.assertEqual(1, self.redisBackend.backend.get.call_count)
        args, kw_args = self.redisBackend.backend.get.call_args
        self.assertIn(key, args[0])
        self.assertIn(self.cacheName, args[0])

    def test_delete(self):
        key = 'Ethyl'
        self.redisBackend.backend.delete = MagicMock()
        self.redisBackend.delete(key)
        self.assertEqual(1, self.redisBackend.backend.delete.call_count)
        args, kw_args = self.redisBackend.backend.delete.call_args
        self.assertIn(key, args[0])
        self.assertIn(self.cacheName, args[0])

    def test_keys(self):
        self.redisBackend.backend.keys = MagicMock(return_value=[])
        self.redisBackend.keys()
        self.assertEqual(1, self.redisBackend.backend.keys.call_count)
        args, kw_args = self.redisBackend.backend.keys.call_args
        self.assertIn(self.cacheName, args[0])


gs/cache/tests/getcache.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
from __future__ import absolute_import, unicode_literals
from unittest import TestCase
from mock import MagicMock
import gs.cache.backends
import gs.cache.getcache
import gs.cache.interfaces


class TestGetCache(TestCase):

    def setUp(self):
        gs.cache.getcache.caches = {}  # Because of the cache of caches

    def test_get_null_cache(self):
        gs.cache.getcache.haveredis = False
        gs.cache.getcache.get_backend = MagicMock(return_value='none')
        r = gs.cache.getcache.get_cache('testcache')
        isCache = gs.cache.interfaces.ICache.providedBy(r)
        self.assertTrue(isCache)
        self.assertIsInstance(r, gs.cache.backends.NullCache)

    def test_get_redis_cache(self):
        gs.cache.getcache.haveredis = True
        gs.cache.getcache.get_backend = MagicMock(return_value='redis')
        gs.cache.getcache.redis = MagicMock()
        r = gs.cache.getcache.get_cache('testcache')
        isCache = gs.cache.interfaces.ICache.providedBy(r)
        self.assertTrue(isCache)
        self.assertIsInstance(r, gs.cache.backends.RedisCache)


gs/cache/tests/test_all.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
from __future__ import absolute_import, unicode_literals
from unittest import TestSuite, main as unittest_main
from gs.cache.tests.getcache import TestGetCache
from gs.cache.tests.backends import TestRedis
from gs.cache.tests.decorator import TestDecorator
testCases = (TestGetCache, TestRedis, TestDecorator, )


def load_tests(loader, tests, pattern):
    suite = TestSuite()
    for testClass in testCases:
        tests = loader.loadTestsFromTestCase(testClass)
        suite.addTests(tests)
    return suite

if __name__ == '__main__':
    unittest_main()


gs/cache/backends.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
###############################################################################
# This code has test cases in tests/test_cache.py.
# Modifications without a supporting test case will be rejected.
#
from __future__ import absolute_import, unicode_literals
from logging import getLogger
log = getLogger('gs.cache')
import sys
if (sys.version_info >= (3, )):
    from pickle import dumps, loads
else:
    from cPickle import dumps, loads  # lint:ok
from threading import RLock
from zope.interface.declarations import implementer
from .interfaces import ICache


@implementer(ICache)
class NullCache(object):

    def __init__(self, backend=None, cache_name=None):
        return

    def set(self, key, object, expiry=None):
        """ Don't set anything, since we don't actually cache. """

    def has_key(self, key):
        """ Since we don't cache, we don't have any keys. """
        return False

    def get(self, key):
        return None

    def keys(self):
        return []

    def delete(self, key):
        return

    def clear(self):
        return


@implementer(ICache)
class RedisCache(object):
    __thread_lock = RLock()

    def __init__(self, backend, cache_name):
        self.backend = backend
        self.cache_name = cache_name

    def set(self, key, value, expiry=None):
        try:
            if not self.__thread_lock.acquire(False):
                m = "Cache ({0}), not adding object ({1}) to cache, would "\
                    "have required blocking".format(self.cache_name, key)
                log.info(m)
                return False
            fullKey = self.cache_name + '%' + key
            self.backend.set(fullKey, dumps(value))
            if expiry:
                self.backend.expire(fullKey, expiry)
        finally:
            try:
                self.__thread_lock.release()
            except:
                pass
        return True

    def has_key(self, key):
        fullKey = self.cache_name + '%' + key

        result = False
        if self.backend.keys(fullKey):
            result = True

        return result

    def get(self, key):
        fullKey = self.cache_name + '%' + key
        result = self.backend.get(fullKey)

        obj = None
        if result:
            obj = loads(result)
        return obj

    def delete(self, key):
        fullKey = self.cache_name + '%' + key
        self.backend.delete(fullKey)

    def keys(self):
        offset = len(self.cache_name) + 1
        outkeys = []
        search_string = self.cache_name + '%*'
        for key in self.backend.keys(search_string):
            outkeys.append(key[offset:])

        return outkeys

    def clear(self):
        search_string = self.cache_name + '%*'
        for key in self.backend.keys(search_string):
            self.backend.delete(key)


gs/cache/getcache.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
###############################################################################
#
# This code has test cases in tests/test_cache.py.
# Modifications without a supporting test case will be rejected.
#
from __future__ import absolute_import, unicode_literals
import logging
log = logging.getLogger('gs.cache')
try:
    import redis
    haveredis = True
except ImportError:
    haveredis = False
from gs.config import Config, getInstanceId
from .backends import NullCache, RedisCache


class BackendError(Exception):
    pass


# Yes, we have a cache of caches :-)
caches = {}


def get_backend(instance_id=None):
    '''Get the cache configuration information.'''
    config = Config(instance_id)
    config.set_schema('cache', {'backend': str, 'hostname': str, 'port': int})
    cache_config = config.get('cache')

    retval = cache_config.get('backend', 'none')
    return retval


def get_cache(cache_name, instance_id=None):
    '''Get the cache for the current instance.'''
    if not instance_id:
        instance_id = getInstanceId()

    # prepend the cache_name with the instance_id to make it unique across
    # instances
    cache_name = instance_id + "::" + cache_name

    # Add the cache to the cache of caches.
    if cache_name in caches:
        return caches[cache_name]

    backend = get_backend(instance_id)

    # --=mpj17=-- TODO: Replace with named adaptors
    if backend == 'redis' and haveredis:
        # TODO: Get the host, port and db from the config file.
        redisCache = redis.StrictRedis(host='localhost', port=6379, db=0)
        caches[cache_name] = RedisCache(redisCache, cache_name)
    elif backend == 'none':
        caches[cache_name] = NullCache()
    else:
        raise BackendError('No cache backend implemented for "%s"' % backend)

    m = 'Intialising cache "{0}" using backend "{1}".'
    msg = m.format(cache_name, backend)
    log.info(msg)

    return caches[cache_name]


gs/cache/decerator.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
###############################################################################
from .getcache import get_cache


def cache(cachename, cachekeyfunc, expiry=None):
    # complex as it looks, this is the main cache decorator
    def cache_decorator(f):
        def do_cache(*args):
            cache = get_cache(cachename)
            cache_key = cachekeyfunc(*args)
            result = cache.get(cache_key)
            if not result:
                result = f(*args)
                cache.set(cache_key, result, expiry)
            return result  # do_cache
        return do_cache  # cache_decorator
    return cache_decorator  # simplecache


gs/cache/decorator.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
###############################################################################
from __future__ import absolute_import, unicode_literals
from .getcache import get_cache


def cache(cachename, cachekeyfunc, expiry=None):
    # complex as it looks, this is the main cache decorator
    def cache_decorator(f):
        def do_cache(*args):
            cache = get_cache(cachename)
            cache_key = cachekeyfunc(*args)
            result = cache.get(cache_key)
            if not result:
                result = f(*args)
                cache.set(cache_key, result, expiry)
            return result  # do_cache
        return do_cache  # cache_decorator
    return cache_decorator  # simplecache


gs/cache/interfaces.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
###############################################################################
from __future__ import absolute_import, unicode_literals
from zope.interface import Interface


class ICache(Interface):

    def set(key, object):
        """ Add an object to the cache.

        """

    def has_key(key):
        """ Check to see if an object is in the cache.

        """

    def get(key):
        """ Get an object from the cache by key.

        """

    def remove(key):
        """ Remove an object from the cache by key.

        """

    def clear():
        """Clear all instances from a cache
        """


gs/cache/cache.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
###############################################################################
#
# This code has test cases in tests/test_cache.py.
# Modifications without a supporting test case will be rejected.
#

from gs.config import Config, getInstanceId
from gs.cache.backends import NullCache, RedisCache
import logging
log = logging.getLogger('gs.cache')

try:
    import redis
    haveredis = True
except ImportError:
    haveredis = False


class BackendError(Exception):
    pass


caches = {}


def get_cache(cache_name, instance_id=None):
    if not instance_id:
        instance_id = getInstanceId()

    # prepend the cache_name with the instance_id to make it unique across
    # instances
    cache_name = instance_id + "::" + cache_name

    # yes, we have a cache of caches :-)
    if cache_name in caches:
        return caches[cache_name]

    config = Config(instance_id)
    config.set_schema('cache', {'backend': str, 'hostname': str, 'port': int})
    cache_config = config.get('cache')

    backend = cache_config.get('backend', 'none')

    if backend == 'redis' and haveredis:
        redisCache = redis.StrictRedis(host='localhost', port=6379, db=0)
        caches[cache_name] = RedisCache(redisCache, cache_name)
    elif backend == 'none':
        caches[cache_name] = NullCache()
    else:
        raise BackendError(u'No cache backend implemented for "%s"' % backend)

    log.info(u'Intialising cache "%s" using backend "%s"' % (cache_name,
                                                                backend))

    return caches[cache_name]


def cache(cachename, cachekeyfunc, expiry=None):
    # complex as it looks, this is the main cache decorator
    def cache_decorator(f):
        def do_cache(*args):
            cache = get_cache(cachename)
            cache_key = cachekeyfunc(*args)
            result = cache.get(cache_key)
            if not result:
                result = f(*args)
                cache.set(cache_key, result, expiry)
            return result  # do_cache
        return do_cache  # cache_decorator
    return cache_decorator  # simplecache
